### PR TITLE
[mypyc] Add 'filter' primitive type

### PR DESCRIPTION
**Pull request by Lynskylate**
### Description

Related   mypyc/mypyc#823

## Test Plan
Add basic test for the primitive

### CODE REVIEW

File: mypyc/test-data/irbuild-basic.test [hunk 1]
```diff
@@ -3649,3 +3649,76 @@ L0:
...
+def filter_object()->None:
+    f = filter(filter_func, range(10))
+    sum = 0
+    for i in f:
+        sum += i
+
+[out]
+def filter_func(x):                           
+    x, r0 :: int                              
+    r1 :: int64
```
Comment (TH3CHARLie) [inline mypyc/test-data/irbuild-basic.test:3666]: `int64` -> `native_int` to fix the CI on 32bit platform

**Review by TH3CHARLie — COMMENTED** [review]
Thanks for the PR! I have a few minor nits. Another important thing to have is the performance benchmark. Can you write a small benchmark to verify the speedup? You can use https://github.com/mypyc/mypyc-benchmarks/tree/master/microbenchmarks as the testing framework

**Comment by JukkaL** [comment]
Hmm this doesn't quite address mypyc/mypyc#823. What we'd want to do with calls to `filter` is to turn them into for loops. Adding a primitive type object is probably not going to be of much benefit. However, if you can demonstrate a significant performance increase, we could consider a simplified version of this that doesn't add a primitive type. Instead, perhaps you can just include the `load_address_op` definition to look up the type object, for equivalent gains. `filter` is hardly ever used as a type, so making it a primitive type is not going to help and just adds extra code that needs to be maintained.

File: mypyc/ir/rtypes.py [hunk 2]
```diff
@@ -324,6 +324,8 @@ def __hash__(self) -> int:
 tuple_rprimitive = RPrimitive('builtins.tuple', is_unboxed=False,
                               is_refcounted=True)  # type: Final
 
+filter_rpimitive = RPrimitive('builtins.filter', is_unboxed=False,
+                              is_refcounted=True)  # type: Final
```
[the remark this answers is not in the document]
Comment (Lynskylate) [reply mypyc/ir/rtypes.py:329]: Done

File: mypyc/ir/rtypes.py [hunk 3]
```diff
@@ -401,6 +403,9 @@ def is_tuple_rprimitive(rtype: RType) -> bool:
     return isinstance(rtype, RPrimitive) and rtype.name == 'builtins.tuple'
 
 
+def is_filter_rprimitive(rtype: RType) -> bool:
+    return isinstance(rtype, RPrimitive) and rtype.name == 'builtins.filter'
+
```
[the remark this answers is not in the document]
Comment (Lynskylate) [reply mypyc/ir/rtypes.py:409]: Done

**Comment by Lynskylate** [comment]
I tried to add a test to the filter in mypyc-benchmark, the code is as follows.
```python
@benchmark
def filter_builtin() -> None:
    a = [i for i in range(1000 * 1000) if i % 5 == 0]


    n = sum(filter(lambda i: i % 2 == 0, a))

    str_a = list(map(str, a))
    filted_list = [] # type: List[Iterable[str]]
    sum_list = [] # type: List[int]
    for i in range(10):
        filted_list.append(filter(lambda x: x.startswith(str(i)), str_a))

    for i in range(10):
        sum_i = 0
        for num in filted_list[i]:
            sum_i += int(num)
        sum_list.append(sum_i)

    assert n == 49999500000, n
```

Before add filter Primitive:
compiled:    0.223820s (avg of 5 iterations; stdev 0.4%)
compiled:    0.223894s (avg of 5 iterations; stdev 0.78%)


After add filter Primitive:
compiled:    0.221474s (avg of 5 iterations; stdev 0.33%)


This change does not significantly improve performance， I will try another way to optimize the filter with little change

**Comment by TH3CHARLie** [comment]
the results prove jukka's theory so you should try to use a specializer to transform `filter` into loops.

**Comment by 97littleleaf11** [comment]
I am going to close this due to the insignificant improvement. Still thanks for your contribution!

**Pull request closed without merging**